rust/src/memcap.rs

```diff
@@ -0,0 +1,257 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+const BASE: isize = 2;
+
+/// enum to store currently allowed memory units in config,
+/// can be easily expanded for more units
+#[derive(Debug, Eq, PartialEq)]
+pub enum MemUnit {
+    B       = BASE.pow(0),
+    KB      = BASE.pow(10),
+    MB      = BASE.pow(20),
+    GB      = BASE.pow(30),
+    Unknown = 0,
+ }
+
+/// Helper function to retrieve memory unit from a string slice
+///
+/// Return value:
+///     MemUnit enum type
+///
+/// # Arguments
+///
+/// * `unit` - A string slice possibly containg memory unit
+///
+/// # Examples
+/// ```
+/// use suricata_rust::memcap::{self, MemUnit};
+///
+/// let unit = "KB";
+/// assert_eq!(MemUnit::KB, memcap::get_memunit(unit));
+///
+/// let unit = "mb";
+/// assert_eq!(MemUnit::MB, memcap::get_memunit(unit));
+///
+/// let unit = "gB";
+/// assert_eq!(MemUnit::GB, memcap::get_memunit(unit));
+/// ```
+pub fn get_memunit(unit: &str) -> MemUnit {
+    let unit = &unit.to_lowercase()[..];
+    match unit {
+        "b"     => { MemUnit::B }
+        "kb"    => { MemUnit::KB }
+        "mb"    => { MemUnit::MB }
+        "gb"    => { MemUnit::GB }
+        _       => { MemUnit::Unknown }
+    }
+}
+
+/// Parses memory units from human readable form to machine readable
+///
+/// Return value:
+///     Result => Ok(f64)
+///            => Err(error string)
+///
+/// # Arguments
+///
+/// * `arg` - A string slice that holds the value parsed from the config
+///
+/// # Examples
+///
+/// ```
+/// use suricata_rust::memcap::{self, MemUnit};
+///
+/// let s = "5kb";
+/// let res = (5 * 1024) as f64;
+/// assert_eq!(Ok(res), memcap::parse(s));
+///
+/// let s = "10";
+/// let res = 10 as f64;
+/// assert_eq!(Ok(res), memcap::parse(s));
+///
+/// let s = "4hb";
+/// assert_eq!(true, memcap::parse(s).is_err());
+/// ```
+pub fn parse(arg: &str) -> Result<f64, &'static str> {
+    let arg = arg.trim();
+    let mut val: &str = "";
+    let mut unit: &str = "";
+    let arg_vec: Vec<&str> = arg.split_whitespace().collect();
+    let arg_vec_len = arg_vec.len();
+    if arg_vec_len > 2 {
+        return Err("Too many whitespaces");
+    } else if arg_vec_len == 1 {
+        val = arg.split(|c: char| c.is_alphabetic()).collect::<Vec<_>>()[0];
+        let unit_vec = arg.split(|c: char| c.is_numeric()).collect::<Vec<_>>();
+        unit = unit_vec.last().unwrap();
+    } else if arg_vec_len == 2 {
+        val = arg_vec[0];
+        unit = arg_vec[1];
+    }
+    val = val.trim();
+    unit = unit.trim();
+    if unit.is_empty() {
+        unit = "B";
+    }
```

Comment:
Any thoughts on using nom for this, it would look something like:
```
        use nom::{
            character::complete::{alpha0, digit1, multispace0},
            sequence::preceded,
            sequence::tuple,
        };
        let parser = tuple((preceded(multispace0, digit1), preceded(multispace0, alpha0)));
        let r: nom::IResult<&str, (&str, &str)> = parser("100  kb");
        if let Ok(r) = r {
            let val = r.1.0;
            let unit = r.1.1;
        }
```
(@victorjulien @catenacyber )

Reply:
hmm I feel ultra stupid rn. Did not even think about nom!

Reply:
We're not really used to using it outside protocols parsers, but this could be a good place for it.

Reply:
This does seem to be working well to extract the values. Thank you!

Reply:
Is it OK for a value `5 1kb` to be parsed as `5B` only? That's one case I'm unable to make it work. I was earlier erroring out on that.